### PR TITLE
Bug 1952268: Increase inertia duration for the EtcdMembersDegraded condition

### PR DESCRIPTION
**Pull request by retroflexer**
It appears rebooting a node and getting it to the point where the kubelet is running static pods can take more than the default 2 minutes. This causes etcd operator to set Degraded=True on healthy machine-config node reboots. 

I am guessing a 5 minute duration is sufficient, but it is the guesswork. I am not sure if there is any slower hardware that would take more than 5 minutes to reboot and restart the static pods.

### CODE REVIEW

**Review by wking — APPROVED** [review]

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -177,7 +178,14 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 		operatorClient,
 		versionRecorder,
 		controllerContext.EventRecorder,
-	)
+	).WithDegradedInertia(status.MustNewInertia(
+		2*time.Minute,
+		status.InertiaCondition{
+			ConditionTypeMatcher: regexp.MustCompile("^EtcdMembersDegraded$"),
```
Comment (lilic) [inline pkg/operator/starter.go:184]: Side note: If library-go folks are onboard ideally, the `InertiaCondition` should have something like `ExactyTypeMatcher`, as we here want the exact match anyways and would say us not needing to call regex. Maybe for a future improvement, not blocking at all.

Comment (retroflexer) [reply pkg/operator/starter.go:184]: I agree. It makes code more readable and easier to maintain. But supporting regex to cover multiple conditions such as ("^(EtcdMembers|EtcdCertSigner|ClusterMember)ControllerDegraded$") is useful too.

**Review by hexfusion — COMMENTED** [review]
I think this would solve the degraded status problem but I wonder if we should consider the root cause of the invalid reporting. etcd health checks are an important part of observability so ensuring health checks are accurately reporting is important. I would expect that degraded status from a "known condition" should preclude an observable "expected condition" by the operator.

tl;dr, when a node goes down for reboot during upgrade is there a status that could be observed telling us "node is down " failed health check is AsExpected true? If that still causes an issue then perhaps we event failed health check but not as a degraded status.

Let's explore these questions a bit and talk about the trade-offs?

https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/etcdmemberscontroller/etcdmemberscontroller.go#L67

**Comment by wking** [comment]
> when a node goes down for reboot during upgrade is there a status that could be observed telling us "node is down " failed health check is AsExpected true?

You could look at the node's `Ready` condition, but if a node is legitimately down for "too long" you still want to complain, right?  I guess you could use a `Ready` node to say you didn't want to bump inertia, to get lower-latency complaints there.

**Comment by wking** [comment]
Also, from the bug's [original job][1]:

```
Apr 21 14:28:20.460 - 112s  E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-243-65.us-east-2.compute.internal is unhealthy
Apr 21 14:34:21.216 - 68s   E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-156-170.us-east-2.compute.internal is unhealthy
Apr 21 14:39:43.199 - 45s   E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-177-210.us-east-2.compute.internal is unhealthy
```

But from [this PR's presubmit][2]:

```
Apr 28 04:49:38.151 - 3s    E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-131-175.us-west-2.compute.internal is unhealthy\nNodeControllerDegraded: The master nodes not ready: node "ip-10-0-131-175.us-west-2.compute.internal" not ready since 2021-04-28 04:47:38 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)
Apr 28 04:59:24.540 - 3s    E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-217-171.us-west-2.compute.internal is unhealthy\nNodeControllerDegraded: The master nodes not ready: node "ip-10-0-217-171.us-west-2.compute.internal" not ready since 2021-04-28 04:57:23 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)
```

So maybe `NodeControllerDegraded` needs an inertia bump too?

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.8-upgrade-from-stable-4.7-e2e-aws-upgrade/1384851693719523328
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-agnostic-upgrade/1387244034145128448

**Comment by hexfusion** [comment]
> You could look at the node's Ready condition, but if a node is legitimately down for "too long" you still want to complain, right?

It feels "broken" that etcd-operator must use a bornderline arbitrary time duration estimation as a condition for cluster sanity during upgrade. These are health checks if a node is up and etcd container is down we want to report that degraded status immediatly to assist in triage. A single source of truth should define this unexpected condition `NodeRebootDurationWarning`? during upgrade (CVO/node?) then the operators can react to that? etcd should not reason about node availability being AsExpected (true|false). Maybe we merge this then come up with a better plan for observability I just want to have this conversation.

> I would expect that degraded status from a "known condition" should preclude an observable "expected condition" by the operator.

^^ this seems missing?

**Comment by hexfusion** [comment]
> So maybe NodeControllerDegraded needs an inertia bump too?

Yeah, this is sort of my point this is sort of whack-a-mole :) and we will potentaily regress. Where if any controller that gives a degraded status about node condition can observe the condition as expected then we can more accurately cascade/report actual failure.

**Comment by hexfusion** [comment]
Who knows better about the expected status of node than node?

cc @rphillips (not trying to make you more work but I am curious about your thoughts on this)

**Comment by rphillips** [comment]
@hexfusion The scheduler has a default 5m40s duration (node-monitor-grace-period + pod-eviction-timeout, not currently configurable) to see if the Kubelet comes back. If the node is missing after this duration then it's workloads get placed onto other nodes. This protects against network partitions and temporary node outages (like reboots).

One scenario that could extend a reboot might be a fsck. 5 minutes seems like a reasonable setting for now

**Comment by hexfusion** [comment]
[original job:](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.8-upgrade-from-stable-4.7-e2e-aws-upgrade/1384851693719523328)
```
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-agnostic-upgrade/1387244034145128448/artifacts/e2e-agnostic-upgrade/openshift-e2e-test/artifacts/e2e.log | grep 'E clusteroperator/etcd condition/Degraded'
Apr 28 04:49:38.151 E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembers_UnhealthyMembers::NodeController_MasterNodesReady changed: EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-131-175.us-west-2.compute.internal is unhealthy\nNodeControllerDegraded: The master nodes not ready: node "ip-10-0-131-175.us-west-2.compute.internal" not ready since 2021-04-28 04:47:38 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)
Apr 28 04:49:38.151 - 3s    E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-131-175.us-west-2.compute.internal is unhealthy\nNodeControllerDegraded: The master nodes not ready: node "ip-10-0-131-175.us-west-2.compute.internal" not ready since 2021-04-28 04:47:38 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)
Apr 28 04:59:24.540 E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembers_UnhealthyMembers::NodeController_MasterNodesReady changed: EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-217-171.us-west-2.compute.internal is unhealthy\nNodeControllerDegraded: The master nodes not ready: node "ip-10-0-217-171.us-west-2.compute.internal" not ready since 2021-04-28 04:57:23 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)
Apr 28 04:59:24.540 - 3s    E clusteroperator/etcd condition/Degraded status/True reason/EtcdMembersDegraded: 2 of 3 members are available, ip-10-0-217-171.us-west-2.compute.internal is unhealthy\nNodeControllerDegraded: The master nodes not ready: node "ip-10-0-217-171.us-west-2.compute.internal" not ready since 2021-04-28 04:57:23 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)
```

[latest job:](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-agnostic-upgrade/1389689359703216128)
```
 $ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-agnostic-upgrade/1389689359703216128/artifacts/e2e-agnostic-upgrade/openshift-e2e-test/artifacts/e2e.log | grep 'E clusteroperator/etcd condition/Degraded'
```

I do see reference to the degraded status in events which is sort of confusing. Is this expected @wking ?

```
Status for clusteroperator/etcd changed: Degraded message changed from "NodeControllerDegraded: The master nodes not ready: node \"ip-10-0-156-74.us-west-2.compute.internal\" not ready since 2021-05-04 22:49:12 +0000 UTC because NodeStatusUnknown (Kubelet stopped posting node status.)\nEtcdMembersDegraded: 2 of 3 members are available, ip-10-0-156-74.us-west-2.compute.internal is unhealthy" to "NodeControllerDegraded: All master nodes are ready\nEtcdMembersDegraded: 2 of 3 members are available, ip-10-0-156-74.us-west-2.compute.internal is unhealthy"
```

**Comment by hexfusion** [comment]
/approve
/lgtm

@wking please unhold if this is as expected.

**Comment by wking** [comment]
Working through the failed presubmits:

* [e2e-agnostic][1]: `aws-ebs-csi-driver-controller-metrics` `TargetDown`, presumably unrelated.
* [e2e-agnostic-upgrade][2]: same thing.  The etcd ClusterOperator never goes `Available=False` or `Degraded=True`, hooray.  We do have some of the `Status for clusteroperator/etcd changed: Degraded message` messages your point out.  That's [expected][3], because we are not hiding the concerning symptoms, we're just staying `Degraded=False` a bit longer while we complain about them in the message.  The e2e-monitor reporting around this is not great; ideally it would include that `False` value in the message.
* [e2e-aws-disruptive][4] failed on unrelated SCC errors.
* [e2e-aws-disruptive-ovn][5] failed on the `aws-ebs-csi-driver-controller-metrics` `TargetDown`.
* [e2e-gcp-disruptive][6] failed on the `aws-ebs-csi-driver-controller-metrics` `TargetDown`.
* [e2e-gcp-disruptive-five-control-plane-replicas][7] failed on the `aws-ebs-csi-driver-controller-metrics` `TargetDown`.
* [e2e-gcp-disruptive-ovn][8] looks like NetworkPolicy exploded.  Also `aws-ebs-csi-driver-controller-metrics` `TargetDown`.
* [gcp-five-control-plane-replicas][9] has `etcdHighNumberOfLeaderChanges` firing.

Looks like y'all will be very happy when [rhbz#1956768][10] lands for the `aws-ebs-csi-driver-controller-metrics` `TargetDown`.  But this PR seems to be working as expected.

/hold cancel

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-agnostic/1389743737088774144
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-agnostic-upgrade/1389743737118134272
[3]: https://github.com/openshift/library-go/blob/7a65fdb398e28782ee1650959a5e0419121e97ae/pkg/operator/status/status_controller_test.go#L157-L171
[4]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-aws-disruptive/1389743737147494400
[5]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-aws-disruptive-ovn/1389743737185243136
[6]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-gcp-disruptive/1389743737227186176
[7]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-gcp-disruptive-five-control-plane-replicas/1389743737256546304
[8]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-gcp-disruptive-ovn/1389743737285906432
[9]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/579/pull-ci-openshift-cluster-etcd-operator-master-e2e-gcp-five-control-plane-replicas/1389743737319460864
[10]: https://bugzilla.redhat.com/show_bug.cgi?id=1956768
